### PR TITLE
Added case study on probabilistic programming and inference

### PR DESCRIPTION
**Pull request by sinaschaefer**
Added the case study `Inference` that contains the contents of my bachelor's thesis `Probabilistic Programming and Programmable Inference in Effekt`.

The content includes Slice Sampling and Metropolis-Hastings algorithms, as well as some examples to show how these algorithms can be applied.

### CODE REVIEW

**Comment by b-studios** [comment]
The tests are failing since a check file is missing. You just need to create a file with the same name that ends with `.check` and contains the expected outputs. You can look into the other case studies to see how this works.

**Comment by b-studios** [comment]
The expected output might be a bit difficult to state since it could rely on `random`. Is it possible to use a different (deterministic) source of randomness for the tests?

File: examples/casestudies/inference.effekt.md [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+    do Weight(prob1 / prob0)
+  }
+  ((result1, trace1), prob1)
+}
+```
+`metropolisStep` is then called in the algorithm implementation
+```
+def metropolisHastingsAlgo[A]{program: () => A / {Sample, Weight}} = {
+  val ((result0, trace0), prob0) = handleWeight{
+    handleRandom{
```
Comment (b-studios) [inline examples/casestudies/inference.effekt.md:278]: Since `handleRandom` is implemented in terms of "global" randomness, the order of handlers here shouldn't matter, right? So we could remove `handleRandom` here and leave randomness uninterpreted.

Comment (sinaschaefer) [reply examples/casestudies/inference.effekt.md:278]: I think it would be possible to move `handleRandom` from this position to the wrapper `metropolisHastings`, but I'm not sure if it is possible to leave the randomness uninterpreted since `Random` is used to generate samples.
In terms of the order of the handlers, since `handleSample` (indirectly) uses the effect `Random`, `handleRandom` has to be applied after applying `handleSample`.

Comment (b-studios) [reply examples/casestudies/inference.effekt.md:278]: I am proposing to move it completely to the outside, that is `main`. This way everything could in principle use randomness, but we can implement it in a deterministic manner.

Comment (b-studios) [reply examples/casestudies/inference.effekt.md:278]: I am proposing to move it completely to the outside, that is `main`. This way everything could in principle use randomness, but we can implement it in a deterministic manner for testing.

**Comment by dvdvgt** [comment]
Thanks for your contribution, @sinaschaefer. In order for the tests to pass, we need to take care of two more things:

1. As you can see [here](https://github.com/effekt-lang/effekt/actions/runs/8860282023/job/24382841865#step:9:802), you need to annotate your extern definition of `pow` with our new ["feature-flag"](https://github.com/effekt-lang/effekt/pull/427). Since we have multiple different backends, one extern definition is ambiguous and compiler can not safely figure out for which backend your extern definition works.
  
    This may all sound horribly complicated, but in the end, you just need to prepend `js` to the right-hand side of the definition:
    ```
    extern pure def pow(x: Double, y: Double): Double = js "Math.pow(x, y)"
    ```
2. Now, we can finally get the CI to run. Since your examples only target the JavaScript backend, we need to ignore your tests when testing the other backends (or add definitions for other backends to the extern definition). For that, you need to add the line `examplesDir / "casestudies" / "inference.effekt.md"` to the files 
    https://github.com/effekt-lang/effekt/blob/master/effekt/jvm/src/test/scala/effekt/LLVMTests.scala
    https://github.com/effekt-lang/effekt/blob/master/effekt/jvm/src/test/scala/effekt/ChezSchemeTests.scala
    https://github.com/effekt-lang/effekt/blob/master/effekt/jvm/src/test/scala/effekt/MLTests.scala
    that is, to all backends for which your extern definitions do not work.

I hope this is helpful, and if not, do not hesitate to shoot us a message.

**Comment by dvdvgt** [comment]
@sinaschaefer, if you'd give me [write permission](https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-access-to-your-personal-repositories/inviting-collaborators-to-a-personal-repository) to your fork, I can happily make these changes for you.

**Comment by dvdvgt** [comment]
@sinaschaefer, it appears you gave me permissions for the wrong repository. I need them for this one: https://github.com/sinaschaefer/effekt. Thanks in advance!

**Comment by sinaschaefer** [comment]
@dvdvgt Oh sorry, now it should be right. Thanks!

**Comment by b-studios** [comment]
@dvdvgt could you round to make test go through?

**Comment by dvdvgt** [comment]
It seems there's no simple solution for this. An generic `Emit[R]` effect is used for printing the results. Therefore, it is not know wether `R` is something that can be rounded before printing. Rounding where `do emit` is envoked does not work either because, there as well, the type of the thing being emited is not know and cannot be assumed to be a double.

I am enclinded to just comment out the parts that cause the failing of the tests due to varying precision of doubles depending on the machine it is run on.

**Comment by b-studios** [comment]
We could handle into a structure like `List[T]` instead of inspecting. This would make testing much easier anyways. 

What do you think @phischu ?

**Review by phischu — APPROVED** [review]

**Comment by phischu** [comment]
Merge it and revise it later.

**Comment by b-studios** [comment]
I agree that we need to merge it soon -- but you know that we will never revise it, if we do not fix it now. Since @dvdvgt is already working on it, we could "quickly" fix it. Otherwise somebody in the future has to wrap their heads around it again and again just to enable these tests.

File: examples/casestudies/inference.effekt.md [hunk 2]
```diff
@@ -309,90 +310,83 @@ def metropolisStep[A](prob0: Probability, trace0: Trace){program: Trace => (A, P
...
+    program()
   }
-
   def loop(result: A, trace: Trace, prob: Probability): Unit / Emit[A] = {
     do emit(result)
-    val ((result1, trace1), prob1)=
-    handleSample{
-      handleRejection{
-        metropolisStepSingleSite(prob, trace){ trace => handleWeight { handleReusingTrace(trace) { program() } } }
+    val ((result1, trace1), prob1)= handleSample {
```
Comment (b-studios) [inline examples/casestudies/inference.effekt.md:380]: Here also the spacing seems off

File: examples/casestudies/inference.effekt.md [hunk 3]
```diff
@@ -295,12 +297,11 @@ def propose(trace: Trace): Trace / Sample = {
...
 
 ```
-def metropolisStep[A](prob0: Probability, trace0: Trace){program: Trace => (A, Probability)} = {
+def metropolisStep[A](prob0: Probability, trace0: Trace) { program: Trace => (A, Probability) } = {
   val trace1 = propose(trace0)
   val (result1, prob1) = program(trace1)
-
-  if(prob1 < prob0){
-    do Weight(prob1 / prob0)
+  if(prob1 < prob0) {
```
[the remark this answers is not in the document]
Comment (dvdvgt) [reply examples/casestudies/inference.effekt.md:303]: I tried to. I guess I missed some

File: examples/casestudies/inference.effekt.md [hunk 4]
```diff
@@ -0,0 +1,617 @@
...
+  with Emit[A] {
+    def emit(element) = {
+      do emit(element);
+      steps = steps - 1;
+      if (steps > 0) { resume(()) }
+    }
+  }
+}
+
+def collect[A,R](n: Int){ program: () => R / Emit[A] }: List[A] = {
```
Comment (b-studios) [inline examples/casestudies/inference.effekt.md:188]: You can delete this, if we didn't use it

File: examples/casestudies/inference.effekt.md [hunk 5]
```diff
@@ -0,0 +1,617 @@
...
+      steps = steps - 1;
+      if (0 < steps) { resume(()) }
+    }
+  }
+  result.reverse
+}
+```
+
+## Rejection Sampling
+The effect `Weight` is also the basis of the rejection handling algorithm
```
Comment (b-studios) [inline examples/casestudies/inference.effekt.md:204]: Here weight should also be lower case

File: examples/casestudies/inference.effekt.md [hunk 6]
```diff
@@ -0,0 +1,617 @@
...
+
+```
+def handleRejection[R] { program: () => R / weight }: R / random = {
+  try { program() }
+  with weight { (prob) =>
+    if (do random() < prob) { resume(()) }
+    else { handleRejection{ program } }
```
Comment (b-studios) [inline examples/casestudies/inference.effekt.md:211]:
```suggestion
    else { handleRejection {program} }
```
Actually in this case I was briefly confused. Maybe calling `program()` here is clearer.

File: examples/casestudies/inference.effekt.md [hunk 7]
```diff
@@ -0,0 +1,617 @@
...
+    with handleObserve
+    do emit(program())
+    loop()
+  }
+  loop()
+}
+```
+
+## Slice Sampling
+
```
Comment (b-studios) [inline examples/casestudies/inference.effekt.md:232]: @phischu is there a sentence of explanation that you could/ would add here?

File: examples/casestudies/inference.effekt.md [hunk 8]
```diff
@@ -0,0 +1,617 @@
...
+}
+```
+
+## Metropolis-Hastings
+
+### Tracing
+A trace records past samples used by an algorithm. The trace also controls where the algorithm draws samples from in the next iterations.
+Constructing traces is possible by handling the effect `Sample`, not with the default handler, but one that draws new samples and records them in a trace.
+
+```
```
Comment (b-studios) [inline examples/casestudies/inference.effekt.md:257]: I would move the definition of type Trace here, where it is explained.

In general, this is a literate programming file: Explanations and definitions should be close to each other.

File: examples/casestudies/inference.effekt.md [hunk 9]
```diff
@@ -0,0 +1,617 @@
...
+  }
+  loop(result, prob)
+}
+```
+
+## Metropolis-Hastings
+
+### Tracing
+A trace records past samples used by an algorithm. The trace also controls where the algorithm draws samples from in the next iterations.
+Constructing traces is possible by handling the effect `Sample`, not with the default handler, but one that draws new samples and records them in a trace.
```
Comment (b-studios) [inline examples/casestudies/inference.effekt.md:255]:
```suggestion
Constructing traces is possible by handling the effect `sample`, not with the default handler, but one that draws new samples and records them in a trace.
```

File: examples/casestudies/inference.effekt.md [hunk 10]
```diff
@@ -0,0 +1,617 @@
...
+
+```
+def handleTracing[R] { program: () => R / sample }: (R, Trace) / sample = {
+  try { (program(), []) }
+  with sample { (dist) =>
+    val d = do sample(dist);
+    val (r, trace) = resume(d);
+    (r, Cons(d, trace))
+  }
+}
```
Comment (b-studios) [inline examples/casestudies/inference.effekt.md:265]: Feel free to ignore this, but why not use mirac6le state, like in other handlers?

```suggestion
  var trace = []
  val resul = try { program() }
  with sample { (dist) =>
    val d = do sample(dist);
    trace = Cons(d, trace)
    resume(d)
  }
  (result, trace.reverse)
}
```

File: examples/casestudies/inference.effekt.md [hunk 11]
```diff
@@ -0,0 +1,617 @@
...
+  def rounded(d: Double) = round(d, 3).show
+  "State(" ++ x.rounded ++ ", " ++ y.rounded ++ ", " ++ vx.rounded ++ ", " ++ vy.rounded ++ ")"
+}
+```
+
+In this example we also define a type alias for `Measurement`
+
+```
+type Measurement = Double
+// movement of the robot in one time intervall
```
Comment (b-studios) [inline examples/casestudies/inference.effekt.md:458]:
```suggestion
// movement of the robot in one time interval
```

maybe quickly run the case stay through a spell checker / LLM

File: examples/casestudies/inference.effekt.md [hunk 12]
```diff
@@ -0,0 +1,617 @@
...
+def move(s: State): State / sample = {
+  s match {
+    case State(x, y, vx, vy) =>
+      val xa = do sample(Gaussian(0.0, 1.0))
+      val ya = do sample(Gaussian(0.0, 1.0))
+      return State(x+vx+xa, y+vy+ya, vx+xa, vy+ya)
+  }
+}
+// measures distance to robot and observes how probable position is given current state
+def measure(s: State, m: Measurement): Unit / observe = {
```
Comment (b-studios) [inline examples/casestudies/inference.effekt.md:468]: This is a literate file, why have a comment here is this can be part of the text?

File: examples/casestudies/inference.effekt.md [hunk 13]
```diff
@@ -0,0 +1,617 @@
...
+```
+record Population(susceptible: Double, infected: Double, recovered: Double)
+
+def show(s: Population): String = {
+  val Population(susceptible, infected, recovered) = s
+  def rounded(d: Double) = round(d, 3).show
+  "Population(" ++ susceptible.rounded ++ ", " ++ infected.rounded ++ ", " ++ recovered.rounded ++ ")"
+}
+
+//multiply s,i,r parts of population with transition probabilitys to get new state of population
```
Comment (b-studios) [inline examples/casestudies/inference.effekt.md:495]: Same as above, also

```suggestion
// multiply s,i,r parts of population with transition probabilitys to get new state of population
```

File: examples/casestudies/inference.effekt.md [hunk 14]
```diff
@@ -0,0 +1,617 @@
...
+  p match {
+    case Population(s, i, r) =>
+      val noise = do sample(Gaussian(0.0, 0.01))
+      val s1 = 0.5 * s - noise
+      val i1 = 0.3 * i + 0.5 * s + 0.1 * r + noise
+      val r1 =  0.9 * r + 0.7 * i
+      return Population(s1, i1, r1)
+  }
+}
+//testing the popultaion(like covid test)
```
Comment (b-studios) [inline examples/casestudies/inference.effekt.md:506]:
```suggestion
// testing the population (like covid test)
```

File: examples/casestudies/inference.effekt.md [hunk 15]
```diff
@@ -0,0 +1,617 @@
...
+def test(p: Population, pr: Double): Unit / observe = {
+  p match {
+    case Population(s, i, r) =>
+      val mean = i
+      val sampleSize = 100.0
+      do observe(pr, Beta(mean, sampleSize))
+  }
+}
+//approximate next state of population based on current state
```
Comment (b-studios) [inline examples/casestudies/inference.effekt.md:515]:
```suggestion
// approximate next state of population based on current state
```

File: examples/casestudies/inference.effekt.md [hunk 16]
```diff
@@ -0,0 +1,617 @@
...
+def step(p0: Population, pr1: Double): Population / { sample, observe } = {
+  val p1 = progression(p0)
+  test(p1, pr1)
+  return p1
+}
+```
+
+Calling the algorithms on the examples.
```
Comment (b-studios) [inline examples/casestudies/inference.effekt.md:523]: I think this sentence could be revised to sum up everything better
```suggestion
Finally, we can use the implemented inference methods to...
```

File: examples/casestudies/inference.effekt.md [hunk 17]
```diff
@@ -0,0 +1,617 @@
...
+# Inference
+This case study shows how we can perform inference over probabilistic models with the help of effects.
+
+---
+
+We require some imports
+
+```
+import list
+extern pure def pow(x: Double, y: Double): Double =
```
Comment (b-studios) [inline examples/casestudies/inference.effekt.md:16]: Why is pow defined here and not in the stdlib?

File: examples/casestudies/inference.effekt.md (hunk 10 above)
Comment (b-studios) [reply examples/casestudies/inference.effekt.md:265]: thanks, typed on my phone and untested

File: examples/casestudies/inference.effekt.md (hunk 6 above)
Comment (dvdvgt) [reply examples/casestudies/inference.effekt.md:211]: This cannot be done. I am not quite sure why, though.

```
/Users/david/Develop/effekt/inference-effekt/out/inference.js:1689
    return handleRejection_2899((weight$capability148573665550262348_20704) => {
                                ^

RangeError: Maximum call stack size exceeded
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1689:33
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1041:22
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1041:22
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1041:22
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1041:22
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1041:22
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1041:22
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1041:22
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1041:22
    at /Users/david/Develop/effekt/inference-effekt/out/inference.js:1041:22

Node.js v22.5.1
[error] Process exited with non-zero exit code 1.
```

File: examples/casestudies/inference.effekt.md (hunk 10 above)
Comment (dvdvgt) [reply examples/casestudies/inference.effekt.md:265]: I am not sure what you mean by "mirac6le state"

**Comment by dvdvgt** [comment]
I did some inference myself and tried to explain most of the parts of the case study, as I found it hard to follow without pre-existing knowledge. The "metropolis hastings section" might still need some more explanations, though.

I tried to add some formalisation to the linear regression example. I am not sure if my interpretation is correct, and I will happily discuss it.

File: libraries/common/effekt.effekt [hunk 18]
```diff
@@ -407,6 +407,19 @@ extern pure def exp(x: Double): Double =
   chez "(exp ${x})"
   ml "Math.exp ${x}"
 
+def pow(base: Double, exponent: Int): Double = {
+  def loop(acc: Double, exponent: Int): Double = {
+    if (exponent == 0) acc
+    else loop(acc * base, exponent - 1)
+  }
+  loop(1.0, exponent)
+}
```
Comment (jiribenes) [inline libraries/common/effekt.effekt:416]: We can also make this slightly faster (logarithmic instead of linear) by using a classic algorithm:
```suggestion
def pow(base: Double, exponent: Int): Double = {
  def loop(base: Double, exponent: Int, acc: Double): Double = {
    if (exponent == 0) acc
    else if (exponent.mod(2) == 0) loop(base * base, exponent / 2, acc)  // even exponent
    else loop(base, exponent - 1, acc * base)  // odd exponent
  }
  loop(base, exponent, 1.0)
}
```

File: examples/casestudies/inference.effekt.md [hunk 19]
```diff
@@ -0,0 +1,693 @@
...
+    -0.13857109526572012,
+    9.9843695780195716 * pow(10.0, 0.0 - 6.0),
+    1.5056327351493116 * pow(10.0, 0.0 - 7.0)]
+  // takes the i-th element of a list
+  def takeith(i: Double, p0: List[Double]): Double = {
+    p0 match {
+      case Nil() => panic("empty list")
+      case Cons(p, ps) =>
+        if(i == 1.0) {p}
+        else {takeith(i - 1.0, ps)}
```
Comment (dvdvgt) [inline examples/casestudies/inference.effekt.md:91]: spaces

File: examples/casestudies/inference.effekt.md [hunk 20]
```diff
@@ -0,0 +1,693 @@
...
+    676.5203681218851,
+    -1259.1392167224028,
+    771.32342877765313,
+    -176.61502916214059,
+    12.507343278686905,
+    -0.13857109526572012,
+    9.9843695780195716 * pow(10.0, 0.0 - 6.0),
+    1.5056327351493116 * pow(10.0, 0.0 - 7.0)]
+  // takes the i-th element of a list
+  def takeith(i: Double, p0: List[Double]): Double = {
```
Comment (dvdvgt) [inline examples/casestudies/inference.effekt.md:86]: This is only used with integer. Use the one from the stdlib.

File: examples/casestudies/inference.effekt.md [hunk 21]
```diff
@@ -0,0 +1,693 @@
...
+    p0 match {
+      case Nil() => panic("empty list")
+      case Cons(p, ps) =>
+        if(i == 1.0) {p}
+        else {takeith(i - 1.0, ps)}
+    }
+  }
+  //lanczos approximation
```
Comment (dvdvgt) [inline examples/casestudies/inference.effekt.md:94]: space

File: examples/casestudies/inference.effekt.md [hunk 22]
```diff
@@ -0,0 +1,693 @@
...
+Currently, we support models with Gaussian, Uniform or Beta distributions. Thus, we need a way to draw from these distributions and to compute the probability density.
+
+```
+def draw(dist: Distribution): Double / random = {
+  dist match {
+    case Gaussian(mean, variance) =>
+      val u1 = do random();
+      val u2 = do random();
+      val sigma = variance
+      //box-muller transform
```
Comment (dvdvgt) [inline examples/casestudies/inference.effekt.md:53]: space

File: examples/casestudies/inference.effekt.md [hunk 23]
```diff
@@ -0,0 +1,693 @@
...
+    current = current * prob;
+    resume(())
+  }
+}
+```
+
+For generating random numbers, we either use the `random` function from stdlib
+
+```
+def handleRandom[R] { program: () => R / random }: R = {
```
Comment (dvdvgt) [inline examples/casestudies/inference.effekt.md:159]: Too many surrounding curly braces

File: examples/casestudies/inference.effekt.md [hunk 24]
```diff
@@ -0,0 +1,693 @@
...
+
+With the effect `emit` it is also possible to limit infinite loops. This allows for flexible numbers of steps to be performed with any algorithm that uses the effect `emit`.
+
+```
+def onEmit[A] { handler: A => Unit } { program: => Any / Emit[A] }: Unit = {
+  try { program(); () }
+  with Emit[A] {
+    def emit(element) = { handler(element); resume(()) }
+  }
+}
```
Comment (dvdvgt) [inline examples/casestudies/inference.effekt.md:190]: Too many curly braces

File: examples/casestudies/inference.effekt.md [hunk 25]
```diff
@@ -0,0 +1,693 @@
...
+  val c = do sample(Gaussian(0.0, 2.0))
+  observations.foreach {
+    case Point(x, y) => do observe(y, Gaussian(m * x + c, 1.0))
+  }
+  return Point(m, c)
+}
+```
+
+Remember from earlier that `handleObserve` uses the `weight` effect operation.
+Therefore, the call `do observe(y, Gaussian(...))` can be understood as approximating $\argmax_{\mathbf{\theta} \in \Theta} P(\mathbf{Y} \mid \mathbf{\theta})$ where $\mathbf{Y}$ are the observations and $\mathbf{\theta}$ are the parameters of a Gaussian given by $m \sim \mathcal{N}(0, 3)$ and $c \sim \mathcal{N}(0, 2)$.
```
Comment (dvdvgt) [inline examples/casestudies/inference.effekt.md:473]: @phischu Please take a look again at this

File: examples/casestudies/inference.effekt.md [hunk 26]
```diff
@@ -0,0 +1,693 @@
...
+```
+
+For simulating the movement of the robot at a given state, we sample accelerations `xa`, `ya` along the x and y axis from a Gaussian distribution.
+By applying the laws of notion and assuming that the difference between each time step is one, we can derive the update equations given the sampled acceleration for the velocity
+$$\begin{align*}a_t = \frac{d v_t}{d t} \Leftrightarrow v_t - v_{t-1} = a_t \Leftrightarrow v_t = v_{t-1} + a_t\end{align*}$$
+and position
+$$v_t = \frac{d x_t}{d t} \Leftrightarrow d x_t = v_t \Leftrightarrow x_t - x_{t-1} = v_t \Leftrightarrow x_t = v_t + x_{t-1} = v_{t-1} + a_t + x_{t-1}.$$
+
+```
+def move(s: State): State / sample = {
```
Comment (dvdvgt) [inline examples/casestudies/inference.effekt.md:505]: Too many curly braces

File: examples/casestudies/inference.effekt.md (hunk 25 above)
Comment (phischu) [inline examples/casestudies/inference.effekt.md:473]:
```suggestion
Remember from earlier that `handleObserve` uses the `weight` effect operation.
Therefore, the call `do observe(y, Gaussian(...))` can be understood as approximating $\argmax_{\mathbf{\theta} \in \Theta} P(\mathbf{Y} \mid \mathbf{\theta})$ where $\mathbf{Y}$ are the observations and $\mathbf{\theta}$ are the parameters of a Gaussian given by $m \sim \mathcal{N}(0, 3)$ and $c \sim \mathcal{N}(0, 2)$.
```
```suggestion
Remember from earlier that `handleObserve` uses the `weight` effect operation.
Therefore, the call `do observe(y, Gaussian(...))` can be understood as $P(\mathbf{Y} \mid \mathbf{\theta})$ where $\mathbf{Y}$ are the observations and $\mathbf{\theta}$ are the parameters of a Gaussian given by $m \sim \mathcal{N}(0, 3)$ and $c \sim \mathcal{N}(0, 2)$.
```

**Review by phischu — APPROVED** [review]

**Comment by b-studios** [comment]
It's great that this was finally merged! 💯 

However... with #590 we now support effects like `emit` better. Maybe you could refactor the emit effect to a singleton operation, still?
